File: 2016/fsharp/Day01.fs
namespace AdventOfCode.FSharp.Y2016

// Day 1: No Time for a Taxicab
module Day01 =
    open AdventOfCode.FSharp.Util

    type Turn =
        | Left
        | Right
        | Straight

    let run (input: byte array) output =
        let instructions =
            seq {
                let mutable pos = 0

                while pos < input.Length do
                    let (pos', steps) = parseIntToDelim input (pos + 1) ','B

                    let turn =
                        match input[pos] with
                        | 'L'B -> Left
                        | 'R'B -> Right
                        | c -> failwithf "bad format, invalid turn: %A" (char c)

                    pos <- pos' + 1

                    yield turn, steps
            }
            |> List.ofSeq

        let followDirection (facing, x, y) (turn, steps) =
            let facing' =
                match turn with
                | Right -> (4 + (facing + 1)) % 4
                | Left -> (4 + (facing - 1)) % 4
                | Straight -> facing

            let (x', y') =
                match facing' with
                | 0 -> x, y + steps
                | 1 -> x + steps, y
                | 2 -> x, y - steps
                | 3 -> x - steps, y
                | f -> failwithf "impossible: %A" f

            facing', x', y'

        let (_, x, y) = instructions |> List.fold followDirection (0, 0, 0)

        x + y |> string |> output 1

        let rec findVisitedTwice visited pos instructions =
            match instructions with
            | (turn, steps) :: xs ->
                let (facing', x', y') = followDirection pos (turn, 1)

                if visited |> Set.contains (x', y') then
                    Some(x', y')
                else
                    let rest = if steps > 1 then (Straight, steps - 1) :: xs else xs
                    let visited' = visited |> Set.add (x', y')
                    rest |> findVisitedTwice visited' (facing', x', y')
            | _ -> None

        let (vx, vy) = instructions |> findVisitedTwice Set.empty (0, 0, 0) |> Option.get

        vx + vy |> string |> output 2

File: 2016/csharp/Day01.cs
namespace AdventOfCode.CSharp.Y2016;

public static class Day01
{
    public static void Run(byte[] input, Action<int, string> output)
    {
        var n = Encoding.UTF8
            .GetString(input)
            .Split('\n')
            .First()
            .Split(',')
            .Select(s => s.Trim())
            .ToList();

        int x = 0;
        int y = 0;

        int ff = 0;
        for (var i = 0; i < n.Count; ++i)
        {
            var v = n[i];
            var d = int.Parse(v[1].ToString());
            var f = v[0];
            if (f == 'R')
            {
                ff += 1;
            }
            else
            {
                ff -= 1;
            }
            ff = ff % 4;

            switch (ff)
            {
                case 0: y += 1; break;
                case 1: x += 1; break;
                case 2: y -= 1; break;
                case 3: x -= 1; break;
            }
        }

        output(1, (x + y).ToString());
    }
}